psi4/src/psi4/libfock/jk.h

```diff
@@ -1193,94 +1182,134 @@ class PSI_API MemDFJK : public JK {
-    size_t memory_estimate() override;
-
-    // => Required Algorithm-Specific Methods <= //
+class SplitJKBase {
+  protected:
+   /// Number of threads to use in the calculation
+   int nthread_;
+   /// Options object
+   Options& options_;
+   /// Print flag, defaults to 1
+   int print_;
+   /// Debug flag, defaults to 0
+   int debug_;
+   /// Bench flag, defaults to 0
+   int bench_;
+
+   /// Whether or not the D matrix is Hermitian
+   bool lr_symmetric_;
+   /// Whether or not to use early screening
+   bool early_screening_;
+
+   /// Integral objects (one per thread)
+   std::vector<std::shared_ptr<TwoBodyAOInt>> ints_;
+   /// The primary basis set
+   std::shared_ptr<BasisSet> primary_;
+   /// Builds the integrals for the J (or K) build class
+   virtual void build_ints() = 0;
+
+  public:
+   /**
+    * @brief Construct a new JK algo object
+    * All Split J/K algorithms extend from this class
+    * 
+    * @param primary The primary basis set used in the J/K matrix computation
+    * @param options The options object
+    */
+   SplitJKBase(std::shared_ptr<BasisSet> primary, Options& options);
 
-    /// Do we need to backtransform to C1 under the hood?
-    bool C1() const override { return true; }
-    /// Setup integrals, files, etc
-    void preiterations() override;
-    /// Compute J/K for current C/D
-    void compute_JK() override;
-    /// Delete integrals, files, etc
-    void postiterations() override;
+   /**
+    * @brief Builds the J/K matrix according to the class algorithm
+    * 
+    * @param D The list of D matrices
+    * @param X The list of J (or K) matrices, depending on algorithm
+    */
+   virtual void build_G_component(const std::vector<SharedMatrix>& D, std::vector<SharedMatrix>& X) = 0;
 
-    /// Build the coulomb (J) matrix
-    void build_J(std::vector<std::shared_ptr<Matrix> >& D,
-                 std::vector<std::shared_ptr<Matrix> >& J);
+   /**
+    * @brief Prints the specific algorithm information for the J/K build
+    */
+   virtual void print_header() = 0;
 
-    /// Build the exchange (K) matrix
-    void build_K(std::vector<std::shared_ptr<Matrix> >& D,
-                 std::vector<std::shared_ptr<Matrix> >& K);
+   /**
+    * @brief Set the early screening variable
+    * 
+    */
+   void set_early_screening(bool early_screening) { early_screening_ = early_screening; }
 
-    /// Common initialization
-    void common_init();
+   /**
+    * @brief Set the lr_symmetric_ variable
+    * 
+    */
+   void set_lr_symmetric(bool lr_symmetric) { lr_symmetric_ = lr_symmetric; }
 
-   public:
-    // => Constructors < = //
+};
 
-    /**
+/**
+ * Class CompositeJK
+ *
+ * JK implementation allowing mixing and matching of different J and K algorithms
+ * (such as Direct-DF J builds, and LinK builds)
+ */
+class PSI_API CompositeJK : public JK {
+  protected:
+   /// Auxiliary basis set used in DF algorithms
+   std::shared_ptr<BasisSet> auxiliary_;
```

Comment:
Quick question: does the CompositeJK class need a member variable for the auxiliary basis set? From my first glance, it seems that auxiliary_ in CompositeJK is only used as a function argument for construction of the Direct DF-J J algorithm in the CompositeJK constructor, at which point you could just use the auxiliary argument from the CompositeJK constructor itself.

Reply:
Not needed. Will make the change for the sake of code cleanliness

Reply:
Actually, based on the `common_init()` convention, it would be needed, unfortunately.

Reply:
Ahhhhhh, yes, I see what you mean, since the common_init function doesn't take function arguments. A bit of a regrettable consequence, but probably better to keep the convention here.